Refs #30331 - refctor nav header to pf4

MariaAga opened the pull request:
Depends on: ~~https://github.com/theforeman/foreman/pull/7804~~ ~~https://github.com/theforeman/foreman/pull/7778~~ ~~https://github.com/theforeman/foreman/pull/7788~~ ~~https://github.com/theforeman/foreman/pull/7861~~
Changes header breakpoint to `1002px`- then the menus are moved to the side nav (was `767px`)
![Screenshot from 2020-07-07 14-55-43](https://user-images.githubusercontent.com/30431079/86777549-f7cd2980-c061-11ea-9620-fae53a2ee5d8.png)

Foreman special context selector design suggestion-
![Screenshot from 2020-08-13 13-38-12](https://user-images.githubusercontent.com/30431079/90133176-b114d280-dd77-11ea-80f9-2d6e02211fdf.png)

MariaAga commented on webpack/assets/javascripts/react_app/components/Layout/LayoutHelper.js:
```diff
@@ -135,3 +136,107 @@ const createLocationItem = locations => {
+    `["navCollapsed","pinnedPath"]`
+  );
+  return !!collapsedState && collapsedState.includes('true');
+};
```
leftover to remove

MariaAga commented:
Moved the layout prop refactor to a separate pr: https://github.com/theforeman/foreman/pull/7848

xprazak2 commented:
propTypes were moved to helper, so this needs a rebase now

MariaAga commented:
Rebased

MariaAga commented on webpack/assets/javascripts/react_app/components/Layout/Layout.js:
```diff
@@ -42,33 +41,22 @@ const Layout = ({
-          changeActiveMenu={changeActiveMenu}
-          stopImpersonationUrl={data.stop_impersonation_url}
-        />
+        <header className="pf-c-page__header" id="navbar-header">
```
This is to get PF css, will be removed when the component will use PF4 layout

sharvit commented:
@rohoover

sharvit commented on webpack/assets/javascripts/react_app/components/Layout/components/TaxonomySwitcher/TaxonomyDropdown.scss:
```diff
@@ -0,0 +1,5 @@
+.pf-c-context-selector__menu-list-item a.organization_menuitem,
+.pf-c-context-selector__menu-list-item a.location_menuitem {
```
Is there a way to avoid using the PF class-names and assign our own class-names instead?
So if PF will change there class-names as they did, less chance for breaking changes.

MariaAga replied:
Changed to not include PF class-names

sharvit commented on webpack/assets/javascripts/react_app/components/Layout/__tests__/integration.test.js:
```diff
@@ -20,26 +20,36 @@ describe('Layout integration test', () => {
-
+    const hostsMenuItem = component.find('.secondary-nav-item-pf > a');
+    const locationToggle = component.find(
+      '#location-dropdown .pf-c-context-selector__toggle'
```
Is there a way to avoid using the PF class-names and assign our own class-names instead?
So if PF will change there class-names as they did, less chance for breaking changes.

MariaAga replied:
line 24: We also use it in the css, can we move this issue to the navigation refactor pr? (https://github.com/theforeman/foreman/pull/7809)
line 26: We can't add a class name to the context selector toggle

sharvit reviewed (commented):
Thanks @MariaAga, overall looks good and works well 👍 

Am I the only one who doesn't like the border around the buttons?

Notice the testing errors

MariaAga commented:
Fixed notification drawer, hope this fixes the test :crossed_fingers: - More tests will be fixed
About the border: I'm not sure what over color we can use other than white, because it looks like the toolbar is using all of our blue colors.
Also I think black doesn't look good: (the border in nav will be blue)
![Screenshot from 2020-07-26 15-54-36](https://user-images.githubusercontent.com/30431079/88479521-58ac9b00-cf58-11ea-9dd0-328a5de2365b.png)

sharvit commented:
From one side we should use the pf4 design language but from the other side I think it would look much better without border.
@rohoover

MariaAga commented:
Adding without border screenshot:
pf3:
![Screenshot from 2020-07-27 11-19-34](https://user-images.githubusercontent.com/30431079/88519715-37e95180-cffb-11ea-9964-2c590f30f87c.png)

pf4:
![Screenshot from 2020-07-27 11-18-58](https://user-images.githubusercontent.com/30431079/88519718-391a7e80-cffb-11ea-906a-2a3196494676.png)

MariaAga commented:
Fixed foreman test fails, Katello test fails looks unrelated

ares commented:
FTR I don't like borders either, OTOH after looking at it for a longer time, I got easily used to it. It'd be weird that the user dropdown look differently, then other dropdowns.

MariaAga commented:
![Screenshot from 2020-07-28 12-13-41](https://user-images.githubusercontent.com/30431079/88644571-ccbb8000-d0cb-11ea-96d9-fbac9b134620.png)
We can add a border to user dropdown, but in the PF4 demos they have user dropdowns without a border. 
Also there isn't an official no border Context Selector in PF4 (as it's not really a dropdown), it would be overriding their css
@Rohoover, What do you think we should do here?

MariaAga commented:
New context design suggestion for foreman:
no border, bright blue on hover/ expend
![Screenshot from 2020-08-13 13-38-12](https://user-images.githubusercontent.com/30431079/90125278-70628c80-dd6a-11ea-8116-cfdd417e003c.png)

sharvit commented:
> New context design suggestion for foreman:
> no border, bright blue on hover/ expend
> ![Screenshot from 2020-08-13 13-38-12](https://user-images.githubusercontent.com/30431079/90125278-70628c80-dd6a-11ea-8116-cfdd417e003c.png)

I like it that way 👍 
I think it might look better if the drop-down to take 100% of the top-nav.

Ron-Lavi commented:
> New context design suggestion for foreman:
> no border, bright blue on hover/ expend
> ![Screenshot from 2020-08-13 13-38-12](https://user-images.githubusercontent.com/30431079/90125278-70628c80-dd6a-11ea-8116-cfdd417e003c.png)

Nice work! didn't test it yet, couple of design thoughts
- will the orgs be filtered while typing or only when hitting search button?
- I think it would be better if the lower-border under the filter will be removed
so it will be clearer that the filter is only in that scope, because there are also the orgs above.

Ron-Lavi commented on app/assets/stylesheets/navigation.scss:
```diff
@@ -56,33 +39,21 @@
+    @media (max-width: 1002px) {
+      display: none !important;
+    }
+  }
```
should we move this file into webpack?

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/layout.scss:
```diff
@@ -87,6 +87,14 @@
     }
   }
+
+  .visible-xs-block {
+    display: none !important;
+
+    @media (max-width: 1002px) {
+      display: block !important;
+    }
+  }
```
I wonder why exactly `1002px` ?
basically it should be visible only on `xs` views, right?

MariaAga replied:
On second thought this actually doesn't make sense..
changed to be pf4 break point for the PageHeader Items.
If the user name is too long and the screen is too small, UI will still look broken but it's also happening in PF4 docs..
for example 
![Screenshot from 2020-08-30 19-06-31](https://user-images.githubusercontent.com/30431079/91663858-13016600-eaf4-11ea-8636-741530d8635c.png)

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/notifications/notifications.scss:
```diff
@@ -3,6 +3,8 @@
 #notifications_container {
   position: static;
   margin-top: -1px;
+  color: var(--pf-global--palette--black-1000);
```
I think you can also use something like: `$--pf-global--palette--black-1000`
but maybe you need to import it correctly from `~patternfly`

sharvit replied:
True.
IMO it is better to use the css variables as they provide more value (change in real-time, supported by the browser).
Also, looks like pf4 are transitioning from scss variables to css variables and it might be wise to embrace it.

MariaAga replied:
I didn't find it in the PF docs, can you point it out to me? :)

MariaAga commented:
@sharvit I think PF will disagree, all of their designs (example: [navigation-system design](https://www.patternfly.org/v4/design-guidelines/usage-and-behavior/navigation-system#context-selector)) include top and bottom padding.
Also the padding here comes from the logo so it might be difficult to change

MariaAga commented:
@laviro About the search - it works while typing, you can take a look how it works in PF4: https://www.patternfly.org/v4/documentation/react/components/contextselector
this PR implements it the same. 
About the border: it comes from PF and I think we should keep it, if they decide otherwise in their discussion we will import it from PF (https://github.com/patternfly/patternfly-design/issues/897)

Ron-Lavi commented:
quick UI things I noticed:
- only the first menu item is `active`, while the selected one should be `active` I guess:
![Screenshot (2)](https://user-images.githubusercontent.com/26363699/91730486-97b0ba80-ebae-11ea-9098-1e753596de02.png)

- You can also see in the pic above that I don't have the filter?
I think it worth to show it even when there are small amount of orgs, to let users to know it exists.

- I think the space between the logo and the dropdown is too small? earlier it was bigger.
![Screenshot (3)](https://user-images.githubusercontent.com/26363699/91730456-91224300-ebae-11ea-8868-fb0df22c7a68.png)

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/components/TaxonomySwitcher/TaxonomyDropdown.js:
```diff
@@ -0,0 +1,121 @@
+        onClick: () => {
+          changeTaxonomy({ title: anyTaxonomyText });
+        },
+        className: `${taxonomyType.toLowerCase()}s_clear`,
+      },
+      {
+        title: __(manageTaxonomyText),
+        href: manageTaxonomyURL,
+        className: taxonomyType.toLowerCase(),
```
I saw `taxonomyType.toLowerCase()` being called on several places,
can you set a `const` for it?

MariaAga replied:
Great idea, added a const

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/components/TaxonomySwitcher/TaxonomyDropdown.js:
```diff
@@ -0,0 +1,121 @@
+      onSelect={onSelect}
+      onSearchButtonClick={onSearchButtonClick}
+      screenReaderLabel="Selected Taxonomy:"
+      showFilter={taxonomies.length > 6}
```
can we always display it?
or from `taxonomies.length > 1` ?

MariaAga replied:
https://github.com/theforeman/foreman/blob/cbf3a1f127849a8056aa13d772186e290ab19b13/webpack/assets/javascripts/react_app/components/Layout/components/TaxonomyDropdown.js#L67
[link to code](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/Layout/components/TaxonomyDropdown.js#L67)

Why should we change existing behavior?

Ron-Lavi replied:
ok then, let's stick with the 6 items :+1:

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/components/TaxonomySwitcher/TaxonomyDropdown.js:
```diff
@@ -0,0 +1,121 @@
+      className="context-selector"
+      {...props}
+    >
+      {filteredItems.map((taxonomy, i) => (
+        <ContextSelectorItem key={i}>
+          <a
+            id={`select_taxonomy_${taxonomy.title}`}
+            className={`${taxonomyType.toLowerCase()}_menuitem`}
+            href={taxonomy.href}
+            onClick={() => {
+              changeTaxonomy({ title: taxonomy.title, id: taxonomy.id });
+            }}
+          >
+            {taxonomy.title}
```
```suggestion
      {filteredItems.map(({ title, href, id }, i) => (
        <ContextSelectorItem key={i}>
          <a
            id={`select_taxonomy_${title}`}
            className={`${declaredOnTopTaxonomyType}_menuitem`}
            href={href}
            onClick={() => {
              changeTaxonomy({ title, id });
            }}
          >
            {title}
```

MariaAga replied:
awesome, thanks

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/components/TaxonomySwitcher/TaxonomySwitcher.js:
```diff
@@ -0,0 +1,80 @@
+        />
+      </ToolbarItem>
+    )}
+    {isLoading && <Spinner size="md" />}
```
is it `isLoading` for the application loading spinner?
if it's loading of the dropdown's data can you use a `skeleton` ?

MariaAga replied:
The loading is for the page loading and not the dropdowns data. For example, the spinner is used when creating a new host and selecting an architecture

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/components/Toolbar/HeaderToolbar.test.js:
```diff
@@ -0,0 +1,21 @@
+  'render HeaderToolbar': {
+    data: hasTaxonomiesMock.data,
+    currentLocation: hasTaxonomiesMock.currentLocation,
+    changeLocation: () => {},
+    currentOrganization: hasTaxonomiesMock.currentOrganization,
+    changeOrganization: () => {},
+    isLoading: false,
+    changeActiveMenu: () => {},
```
```suggestion
    changeLocation: noop,
    currentOrganization: hasTaxonomiesMock.currentOrganization,
    changeOrganization: noop,
    isLoading: false,
    changeActiveMenu: noop,
```

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/layout.scss:
```diff
@@ -87,6 +87,14 @@
     }
   }
+
+  .visible-xs-block {
+    display: none !important;
+
+    @media (max-width: $pf-global--breakpoint--lg) {
+      display: block !important;
+    }
+  }
```
can you avoid using `!important`?

MariaAga replied:
Changed the class name to not clash with bootstrap

MariaAga commented:
@laviro the 'active' is the item that is focused currently. The outline is there for accessibility so the user will know which item they can click 'enter' on instead of clicking the mouse

MariaAga commented:
![Screenshot from 2020-09-01 14-43-52](https://user-images.githubusercontent.com/30431079/91845314-97bdc280-ec61-11ea-8d75-6f8b5050c3bc.png)
Added padding for the logo

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/components/TaxonomySwitcher/TaxonomySwitcher.js:
```diff
@@ -0,0 +1,80 @@
+    {taxonomiesBool.organizations && (
+      <ToolbarItem>
+        <TaxonomyDropdown
+          taxonomyType="Organization"
```
![Screenshot (7)](https://user-images.githubusercontent.com/26363699/93014178-94a6c880-f5b7-11ea-8500-2da1d51e5ecd.png)

The caret is a quite far from the actual dropdown title, is there a way to move it closer like in the user dropdown?

MariaAga replied:
The width comes from PF `var(--pf-c-context-selector--Width)` which is 250px

Ron-Lavi replied:
I guess that when it was wrapped by a white border it looked connected to the dropdown,
though now you get the feeling that it floated away a bit..

Ron-Lavi replied:
it's clear enough when hovering the dropdown that it belongs to it.. let's leave it that way :+1:

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/components/TaxonomySwitcher/TaxonomyDropdown.js:
```diff
@@ -0,0 +1,122 @@
+        title: __(anyTaxonomyText),
+        href: anyTaxonomyURL,
+        onClick: () => {
+          changeTaxonomy({ title: anyTaxonomyText });
```
While changing taxonomy, can you display the nav loading spinner?

MariaAga replied:
When the taxonomy changes it changes the text immediately and then reloads the page so you get the spinner from the browser, should we add another spinner?

Ron-Lavi reviewed (commented):
Can the click on the dropdown item will also change the org?
currently it happens only when clicking on the text 

![Audits (1)](https://user-images.githubusercontent.com/26363699/93014666-85297e80-f5bb-11ea-9ec4-b6f88fef21f4.gif)

Ron-Lavi commented:
> Can the click on the dropdown item will also change the org?
currently it happens only when clicking on the text

[ContextSelectorItem](https://github.com/theforeman/foreman/pull/7805/files#diff-1a932e67884fd32cdf1e4fa5ce4f9889R76) renders a button, and since we are doing a full page reload using `href` (something that should be improved in the future)
I see why you used the `<a>` tag.. although in this example I would pass the `href` effect through the `onClick` prop to `ContextSelectorItem`, e.g:

Didn't test it, but you get the idea:
```js
onClick={() = { item.onClick(); if(item.href){ window.location.href = item.href } }}
```

MariaAga commented:
@laviro I used your onClick suggestion with 'assign' instead of 'href' and it works great now. Had to use assign as it was easier to test

MariaAga commented:
Fixed ruby tests,
@laviro @sharvit Any more comment?

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/Layout.fixtures.js:
```diff
@@ -190,15 +191,6 @@ const organizations = {
     { id: 2, title: 'org2', href: '/organizations/2-org2/select' },
   ],
-  many_organizations: [
-    { id: 1, title: 'org1', href: '/organizations/1-org1/select' },
```
just making sure, it wasn't used?

MariaAga replied:
This was used to test the `TaxonomyDropdown` search. Now that the search is in pf I removed the test. i'll add tests to check the other functionalities in this component

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/Layout.js:
```diff
@@ -42,34 +41,22 @@ const Layout = ({
+            href={data.root}
+          />
+          <HeaderToolbar
+            data={data}
```
how about:
```suggestion
      {...data}
```

to make props simpler in `HeaderToolbar`

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/__tests__/__snapshots__/Layout.test.js.snap:
```diff
@@ -53,97 +53,175 @@ exports[`Layout rendering renders layout 1`] = `
+                "current_location": "london",
+              },
+              "logo": "/assets/header_logo-c9614c16f2ee399ae9cb7f36ec94b9a26bf8cf9eabaa7fe6099bf80d1f7940db.svg",
+              "menu": Array [
+                Object {
+                  "children": Array [
+                    Object {
+                      "exact": true,
```
I think 2-3 menu items would be enough for the snapshot, it will reduce the file size

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/components/TaxonomySwitcher/TaxonomyDropdown.js:
```diff
@@ -0,0 +1,122 @@
+  const onSelect = () => {
+    setIsOpen(!isOpen);
+  };
+  const onSearchInputChange = (value, event) => {
+    setSearchValue(event.target.value);
+    onSearchButtonClick();
```
the filter doesn't work good as earlier, in some cases, it doesn't show the correct values when the input change :(

MariaAga replied:
In the PF demo they have this code but without ` onSearchButtonClick();` so it only searches when a user clicks. I think I'll change to that so it will be simpler, @laviro what do you think?

Ron-Lavi replied:
I think that for now if we filter upon input change, we should keep doing so,
though if things are not working well, we could stick with the PF4 implementation and open an issue there.

MariaAga replied:
This is not a PF4 issue, apparently the useState set isn't immediate, so I work around that :)

Ron-Lavi reviewed (commented):
Great work @MariaAga, works really well,
although the filter doesn't work well as before onChange,
also, there are some style responsiveness small issues:

![layout_smaller_screens](https://user-images.githubusercontent.com/26363699/97961087-4a71e480-1dbb-11eb-8661-73b4e3ba92d1.png)

![layout_flying_helmet](https://user-images.githubusercontent.com/26363699/97961093-4e056b80-1dbb-11eb-84ec-680a8e7037dc.png)

MariaAga commented:
Thanks @laviro!
Added a new commit to address the requests, will merge once approved.
Changes: 
* Updated the selector search to update in time
* Added times for the search
* removed some of the menu items in tests (so the list in snapshots will be shorter)
* removed unused css (`$switcher-max-height` is not used since we moved the breadcrumbs to PF)
* changed the breakpoint for the header items visibility to be a bit larger but not as large as PF4 XL which is set to 1200 (now its 1062)

The header still might look broken if the username is very long, but there should be enough space for "regular" length usernames

Need to fix ruby tests

Ron-Lavi reviewed (commented):
Thanks for the fixes @MariaAga, almost there!

I saw that when the vertical-nav goes hidden on 767px, the brand is collapsed and still displayed:
![Screenshot (5)](https://user-images.githubusercontent.com/26363699/98543785-a3e68180-229b-11eb-95e5-6cae469cb7be.png)

MariaAga commented:
@laviro Nice catch! added a fix in `HeaderToolbar.scss `. hopefully I will remove it when the navigation will be in PF4

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/__tests__/__snapshots__/LayoutHelper.test.js.snap:
```diff
@@ -42,60 +37,38 @@ Array [
-        "name": "Domains",
-        "type": "item",
-        "url": "/domains",
-      },
-      Object {
-        "name": "Realms",
-        "type": "item",
-        "url": "/realms",
+        "name": undefined,
+        "type": "divider",
```
some `names` went `undefined`, changes in fixtures?

MariaAga replied:
Yeah, I used the undefined menu object instead of using 2 full menu objects 
https://github.com/theforeman/foreman/blob/7f54445542b35ce4df1737a69c5e69fbd30c043a/webpack/assets/javascripts/react_app/components/Layout/Layout.fixtures.js#L162

Ron-Lavi replied:
so it's expected... great :D

Ron-Lavi reviewed (commented):
Everything works great!
only one test snapshot nitpick and LGTM from me :+1:

Ron-Lavi reviewed (dismissed):
Thanks @MariaAga ! LGTM :)

MariaAga commented:
@amirfefer Can you take a look?

MariaAga commented:
@sharvit Rebased.

MariaAga commented:
Changed the pr to use the current taxonomies via context.
Changed the `customContextSelector` to PF4 context selector (Design discussion here: https://github.com/patternfly/patternfly-design/issues/897)
![Screenshot from 2021-02-22 12-05-50](https://user-images.githubusercontent.com/30431079/108712316-43fca980-751f-11eb-9e03-ee54a54ed1c0.png)

Ron-Lavi commented on webpack/assets/javascripts/react_app/Root/Context/ForemanContext.js:
```diff
@@ -10,6 +10,7 @@ const useForemanMetadata = () => useForemanContext().metadata;
 export const useForemanDocUrl = () => useForemanMetadata().docUrl;
-export const useForemanOrganization = () => useForemanMetadata().organization;
-export const useForemanLocation = () => useForemanMetadata().location;
+export const useForemanOrganization = () =>
+  useForemanMetadata().organization?.title;
+export const useForemanLocation = () => useForemanMetadata().location?.title;
```
In case someone else will want to use something other than the `title`
it would be better to return only the `org/loc` and take the needed value inside your component.

Ron-Lavi reviewed (commented):
Overall looks great and works as expected,

few comments:
- can you align the `manage organization/location` text to the center?
![Screenshot (67)](https://user-images.githubusercontent.com/26363699/108839330-9f8d6c80-75dd-11eb-81a1-6cd2189c667c.png)

- can you disable clicking on the current organization/location from the list?
![Hosts](https://user-images.githubusercontent.com/26363699/108839515-d6fc1900-75dd-11eb-8e39-43212e911186.gif)

MariaAga commented:
@laviro 

> can you align the manage organization/location text to the center?

I can use a smaller button, but generally that's what pf told us to do with this component (https://github.com/patternfly/patternfly-design/issues/897)

MariaAga commented:
Added a new commit to address the changes
![Screenshot from 2021-02-24 09-56-17](https://user-images.githubusercontent.com/30431079/108967086-b8eaf300-7687-11eb-9858-f4bfbfbabb1c.png)

ezr-ondrej commented on webpack/assets/javascripts/react_app/components/Layout/LayoutReducer.js:
```diff
@@ -37,10 +37,10 @@ export default (state = initialState, action) => {
 
     case LAYOUT_CHANGE_ORG:
-      return state.set('currentOrganization', payload.organization);
+      return state.set('isLoading', true);
```
As mentioned in https://github.com/theforeman/foreman/pull/8255#issuecomment-796599231 IMHO we don't need fancy UI loader here, when this is indicated by browser loading page state.

sharvit replied:
I tend to agree, I see it confusing having 2 loading indicators (browser and in the UI).
IMO we should use them to indicate that the UI is working on sending-receiving data.

Ron-Lavi replied:
I don't think it's too bad to show the loading indicator, we have pages where it takes a few seconds to do a full-refresh,
and until that happens, the data which is visible is from the prior org/loc...
ideally, I would show some loading skeleton for each page, but that should be on another PR,
here's an example of how it looks now:

![Untitled_ Mar 14, 2021 5_01 PM](https://user-images.githubusercontent.com/26363699/111073626-1be1e400-84e8-11eb-940a-dc574c345cf6.gif)

ezr-ondrej replied:
I believe there shouldn't be redux reducer and action for this even if we want to do so :)
Redux is for maintaining the data. I disagree with the point of "we see data from the previous taxonomy", we do, because the page didn't reload. The UI stays correctly on the old taxonomy, so it is still showing the old data and is in reload page state.

ezr-ondrej reviewed (commented):
I believe we should merge https://github.com/theforeman/foreman/pull/8255 first and rebase on top of that and it would even simplify this PR.

Ron-Lavi commented on webpack/assets/javascripts/react_app/common/variables.scss:
```diff
@@ -0,0 +1,3 @@
+@import '~@theforeman/vendor/scss/variables';
+
+$header-max-width: calc(#{$pf-global--breakpoint--lg} + 70px);
```
can you add a `TODO` and open a redmine issue to move this into `@theforeman/vendor/scss/variables` package?
we won't need an extra file and anyone could import it from one known location,

if this rule is applicable only for the nav layout, you can also set it in the layout scss file

MariaAga replied:
https://projects.theforeman.org/issues/32096

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/components/ImpersonateIcon/ImpersonateIcon.js:
```diff
@@ -25,11 +25,9 @@ const ImpersonateIcon = props => {
-          </span>
-        </li>
+        <span className="nav-item-iconic" onClick={toggleModal}>
+          <Icon name="eye small" className="blink-image" />
```
no need for the `avatar` icon anymore? :)
also, can we use the new PF4 Icon?

MariaAga replied:
Using pf4 icon now

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/components/TaxonomySwitcher/TaxonomyDropdown.js:
```diff
@@ -0,0 +1,137 @@
+  const [isOpen, setIsOpen] = useState(false);
+  const [filteredItems, setFilteredItems] = useState(taxonomies);
+
+  const onSearchButtonClick = useCallback(() => {
```
Nice usage of callback! :+1:

Ron-Lavi commented on webpack/assets/javascripts/react_app/components/Layout/Layout.js:
```diff
@@ -42,27 +39,20 @@ const Layout = ({
-          stopImpersonationUrl={data.stop_impersonation_url}
-          instanceTitle={data.instance_title}
-        />
+        <header className="pf-c-page__header pf-c-page" id="navbar-header">
```
Is there a PF4 component with this header? instead of typing it by yourself

MariaAga replied:
There is one, it's in the next pr (https://github.com/theforeman/foreman/pull/7809)
 https://github.com/theforeman/foreman/pull/7809/files#diff-68bb61cb73281bdbbb65e8b2cb6aa72995e43b2e6932a7bbe68bb4e871b17cfdR35

Ron-Lavi reviewed (commented):
Overall looks great, left a few inline comments

Ron-Lavi reviewed (dismissed):
Thanks @MariaAga!! great job with this PR

MariaAga committed:
Refs #30331 - refctor nav header to pf4

ezr-ondrej commented:
[test katello] I thought katello was fixed already, but it is unrelated.

ezr-ondrej reviewed (approved):
Works for me :+1:

sharvit reviewed (approved):
Thanks @MariaAga looks good and works as expected.
